pkg/operator/controller/dns/controller.go

```diff
@@ -588,6 +595,36 @@ func (r *reconciler) createDNSProvider(dnsConfig *configv1.DNS, platformStatus *
 		if err != nil {
 			return nil, fmt.Errorf("failed to create IBM DNS manager: %v", err)
 		}
+	case configv1.AlibabaCloudPlatformType:
+		if platformStatus.AlibabaCloud.Region == "" {
+			return nil, fmt.Errorf("missing region id in platform status")
+		}
+
+		var privateZones []string
+		if dnsConfig.Spec.PrivateZone != nil {
+			privateZones = append(privateZones, dnsConfig.Spec.PrivateZone.ID)
+		}
+
+		cred, err := fetchAlibabaCredentialsIniFromSecret(creds)
+		if err != nil {
+			return nil, err
+		}
+
+		c, ok := cred.(*credentials.AccessKeyCredential)
+		if !ok {
+			return nil, fmt.Errorf("failed to convert the credential to an AccessKeyCredential")
+		}
+
+		provider, err := alidns.NewProvider(alidns.Config{
+			Region:       platformStatus.AlibabaCloud.Region,
+			AccessKeyID:  c.AccessKeyId,
+			AccessSecret: c.AccessKeySecret,
+			PrivateZones: privateZones,
```

Comment:
Does this mean that the client needs to be re-initialized if `dnsConfig.Spec.PrivateZone` changes?  If so, `createDNSProviderIfNeeded` needs to check for this, something like this:
```diff
diff --git a/pkg/operator/controller/dns/controller.go b/pkg/operator/controller/dns/controller.go
index 22c219c5..64ea76af 100644
--- a/pkg/operator/controller/dns/controller.go
+++ b/pkg/operator/controller/dns/controller.go
@@ -122,6 +122,7 @@ type reconciler struct {
 	client           client.Client
 	cache            cache.Cache
 	dnsProvider      dns.Provider
+	dnsConfig        *configv1.DNS
 	infraConfig      *configv1.Infrastructure
 	cloudCredentials *corev1.Secret
 	recorder         record.EventRecorder
@@ -247,13 +248,17 @@ func (r *reconciler) createDNSProviderIfNeeded(dnsConfig *configv1.DNS) error {
 		needUpdate = true
 	}
 
+	if platformStatus.Type == configv1.AlibabaCloudPlatformType && r.dnsConfig == nil || !reflect.DeepEqual(dnsConfig.Spec.PrivateZone, r.dnsConfig.Spec.PrivateZone) {
+		needUpdate = true
+	}
+
 	if needUpdate {
 		dnsProvider, err := r.createDNSProvider(dnsConfig, platformStatus, &infraConfig.Status, creds)
 		if err != nil {
 			return fmt.Errorf("failed to create DNS provider: %v", err)
 		}
 
-		r.dnsProvider, r.infraConfig, r.cloudCredentials = dnsProvider, infraConfig, creds
+		r.dnsProvider, r.dnsConfig, r.infraConfig, r.cloudCredentials = dnsProvider, r.dnsConfig, infraConfig, creds
 	}
 
 	return nil
```

Reply:
Yes, and I considered to move the ID lookup logic to the private zone service, and it finds and caches zoneName to zoneId results when do some operations (Add, Update, Delete) to the private zone. It seems to be a better way to deal with the DNS changing problem.